### PR TITLE
TF generate refactor - Sample

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

This PR is step 3 of the `generate()` refactor plan (see https://github.com/huggingface/transformers/pull/15562) -- it refactors the sample case.

A few notes:
- `_generate_no_beam_search()` was no longer reachable, so it was removed;
- I've tested the changes against the example in `_generate()` and all looks well. There are also a bunch of related tests that are passing after the changes (e.g. pipeline tests). However, the output is non-deterministic. Is there any way I can do stricter testing?
- While inspecting other `generate` files, I came across a few incorrect type hints and corrected them. LMK if you want me to move it to a different PR;

### CODE REVIEW

File: src/transformers/generation_tf_utils.py
Comment (gante) [inline src/transformers/generation_tf_utils.py:1767]: This was moved a few lines below, as it needed to be called after `_expand_inputs_for_generation()` in the sample generate mode

**Review by sgugger — APPROVED** [review]
All looks clean on my side, but @patrickvonplaten will know better :-)

File: docs/source/internal/generation_utils.mdx
Comment (patrickvonplaten) [inline docs/source/internal/generation_utils.mdx:162]: Thanks!

File: src/transformers/generation_flax_logits_process.py
Comment (patrickvonplaten) [inline src/transformers/generation_flax_logits_process.py:97]: Thanks !

File: src/transformers/generation_utils.py
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:559]: Thanks for fixing those

**Comment by gante** [comment]
@patrickvonplaten absolutely agreed, running != being correct. I will add those in a separate PR, and rebase this one as soon as the new stricter tests get merged. Hopefully, no changes will be needed here 🤞

**Comment by gante** [comment]
@patrickvonplaten @Rocketknight1 rebased with the recently merged TF sample generate integration tests (which are passing), with TF logit wrapper tests added on top. We should be ready to merge, but I'd like to get your eyes for explicit approval :)

File: src/transformers/generation_tf_logits_process.py
Comment (Rocketknight1) [inline src/transformers/generation_tf_logits_process.py:146]:
```suggestion
    [`TFLogitsWarper`] that performs top-p, i.e. restricting to top tokens summing to <= prob_cut_off.
```

File: src/transformers/generation_tf_logits_process.py
Comment (Rocketknight1) [inline src/transformers/generation_tf_logits_process.py:139]: Nit: I think you can avoid the `expand_dims` here?
```suggestion
        indices_to_remove = scores < tf.math.top_k(scores, k=top_k)[0][..., -1:]
```

**Review by Rocketknight1 — APPROVED** [review]
Overall, this looks great, and the tests are very solid! I briefly panicked about the use of `top_k` before I remembered that it does sort the outputs by default (I think in the past it didn't..? Maybe I'm misremembering).

File: src/transformers/generation_tf_utils.py
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:1557]: nice!

File: src/transformers/generation_tf_utils.py
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:2046]: indeed!

**Review by patrickvonplaten — APPROVED** [review]
Nothing to add from my side - very clean!

**Comment by patrickvonplaten** [comment]
Before merging could you run a final test that all of the following tests pass in TF:

```
RUN_SLOW=1 pytest -sv \
tests/test_modeling_tf_bart.py \
tests/test_modeling_tf_t5.py \
tests/test_modeling_tf_gpt2.py \
tests/test_modeling_tf_vision_encoder_decoder.py \
tests/test_modeling_tf_encoder_decoder.py \
tests/test_modeling_tf_speech_to_text.py
```

File: src/transformers/generation_tf_logits_process.py
Comment (gante) [reply src/transformers/generation_tf_logits_process.py:139]: (going to merge the suggestion and run the tests 👍  will revert if it fails)

**Comment by gante** [comment]
Tests passing after this commit -- merging as soon as CI gets to green

**Pull request merged**
